src/cpp/session/modules/jobs/ScriptJob.cpp

```diff
@@ -0,0 +1,334 @@
+/*
+ * ScriptJob.cpp
+ *
+ * Copyright (C) 2009-18 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <boost/asio.hpp>
+
+#include <r/RExec.hpp>
+
+#include <core/Algorithm.hpp>
+
+#include <session/SessionModuleContext.hpp>
+#include <session/SessionAsyncRProcess.hpp>
+ 
+#include "ScriptJob.hpp"
+#include "JobsApi.hpp"
+
+#define kProgressDelim  "__"
+#define kProgressStart  kProgressDelim "rs_progress_0" kProgressDelim " " 
+#define kProgressEnd    " " kProgressDelim "rs_progress_1" kProgressDelim
+#define kProgressMiddle " " kProgressDelim " "
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {      
+namespace jobs {
+
+namespace {
+
+class ScriptJob : public async_r::AsyncRProcess 
+{
+public:
+   static boost::shared_ptr<ScriptJob> create(
+         const ScriptLaunchSpec& spec,
+         boost::function<void()> onComplete)
+   {
+      boost::shared_ptr<ScriptJob> pJob(new ScriptJob(spec, onComplete));
+      pJob->start();
+      return pJob;
+   }
+   
+   std::string id()
+   {
+      if (job_)
+         return job_->id();
+      return "";
+   }
+   
+private:
+   ScriptJob(const ScriptLaunchSpec& spec, 
+         boost::function<void()> onComplete):
+      spec_(spec),
+      completed_(false),
+      onComplete_(onComplete)
+   {
+   }
+
+   void start()
+   {
+      Error error;
+      
+      // add the job -- currently idle until we get some content from it
+      job_ = addJob(spec_.path().filename(), "", "", 0, JobIdle, false);
+
+      std::string importRdata = "NULL";
+      std::string exportRdata = "NULL";
+
+      if (spec_.importEnv())
+      {
+         // create temporary file to save/load the data
+         FilePath::tempFilePath(&import_);
+         importRdata = "'" + string_utils::utf8ToSystem(import_.absolutePath()) + "'";
+
+         // prepare the environment for the script by exporting the current env
+         setJobStatus(job_, "Preparing environment");
+         r::exec::RFunction save("save.image");
+         save.addParam("file", import_.absolutePath());
```

Comment:
Does this need `utf8ToSystem()`?

Reply:
No, I don't think we need that when passing parameters directly to R; my understanding is that it's only necessary when preparing system command arguments (we do use it elsewhere in this file for that purpose). Does that align with your understanding?

Reply:
I think the main issue is that the string is passed to R with `"unknown"` encoding, and so R will assume it's in the system default encoding. Since we're calling `save.image()` directly, I think we need to give it a path in the system-native encoding.

When we invoke our own R functions, we often explicitly re-assign the encoding as UTF-8 to work around this; we won't have that opportunity when calling `save.image()` directly.

Reply:
Got it, thanks!